fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/recordrepair/RecordVersionValidator.java

```diff
@@ -57,7 +75,38 @@ public CompletableFuture<RecordValidationResult> validateRecordAsync(final Tuple
     }
 
     @Override
-    public CompletableFuture<Void> repairRecordAsync(final Tuple primaryKey, final CompletableFuture<RecordValidationResult> validationResult) {
-        throw new UnsupportedOperationException("Repair is not yet supported");
+    public CompletableFuture<RecordValidationResult> repairRecordAsync(@Nonnull final RecordValidationResult validationResult) {
+        if (validationResult.isValid()) {
+            // do nothing
+            return CompletableFuture.completedFuture(validationResult.withRepair(RecordValidationResult.REPAIR_NOT_NEEDED));
+        }
+        switch (validationResult.getErrorCode()) {
+            case CODE_RECORD_MISSING_ERROR:
+                // Nothing to do
+                return CompletableFuture.completedFuture(validationResult.withRepair(RecordValidationResult.REPAIR_NOT_NEEDED));
+            case CODE_VERSION_MISSING_ERROR:
+                if (logger.isInfoEnabled()) {
+                    logger.info(KeyValueLogMessage.of("Record repair: Version created",
+                            LogMessageKeys.PRIMARY_KEY, validationResult.getPrimaryKey()));
+                }
+                // Create a new version
+                // Save the record with the existing data and the new version
+                // This uses the WITH_VERSION behavior to force a version even if the metadata says otherwise, since this
+                // is the assumption of the validator
```

Comment:
This means that the caller first has to check the metadata, and determine whether or not versions should be saved, right? Is there a benefit to pushing that check up? We could do it here, but we would need to check it on every valid record.

Reply:
This is now using `DEFAULT` policy since the metadata is assumed to have `isVersionsStored` be `TRUE` (or else we don't check and repair versions at all).